### PR TITLE
Implement Command Parser 2.0 and add ModuleManager

### PR DESCRIPTION
**Pull request by 3m0W33D**
* Major Overhaul to the Command Parser
* Add ModuleManager to manage NusModule
* Each Module is now their own workspace
* Module workspace allows CRUD modules (update not yet done)
* Each Note/Schedule now belong to a module
* JavaDocs for new class 
* Implementing JUnit testing
* Include simple assertion

~~**DO NOT MERGE THIS BRANCH YET**~~

Junit Testing
- [x] ModuleManager
- [x] GoCommand
- [x] WorkspaceCommandParser
- [x] ModuleCommandParser
- [x] AddModuleCommand
- [x] ViewModuleCommand
- [x] DeleteModuleCommand
- [x] ModuleCommand

### CODE REVIEW

**Review by Woolicious98 — APPROVED** [review]
LGTM

File: src/main/java/terminus/Terminus.java [hunk 1]
```diff
@@ -28,7 +29,7 @@
     private String workspace;
 
     private ModuleStorage moduleStorage;
-    private NusModule nusModule;
+    private ModuleManager moduleManager;
 
     private static final String INVALID_ARGUMENT_FORMAT_MESSAGE = "Format: %s";
```
Comment (LouisLouis19) [inline src/main/java/terminus/Terminus.java:34]: Could consider refactoring this to Messages.java

File: src/main/java/terminus/Terminus.java [hunk 2]
```diff
@@ -58,24 +59,24 @@ private void start() {
             this.parser = MainCommandParser.getInstance();
             this.workspace = "";
             this.moduleStorage = new ModuleStorage(DATA_DIRECTORY.resolve(MAIN_JSON));
-
+            this.moduleManager = new ModuleManager();
             TerminusLogger.info("Loading file...");
-            this.nusModule = moduleStorage.loadFile();
+            this.moduleManager = moduleStorage.loadFile();
         } catch (IOException e) {
             TerminusLogger.warning("File loading has failed.", e.fillInStackTrace());
```
Comment (LouisLouis19) [inline src/main/java/terminus/Terminus.java:66]: You didn't make these changes, but we could consider using constants for these messages :)

File: src/main/java/terminus/command/GoCommand.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import terminus.exception.InvalidArgumentException;
+import terminus.exception.InvalidCommandException;
+import terminus.module.ModuleManager;
+import terminus.module.NusModule;
+import terminus.parser.ModuleWorkspaceCommandParser;
+import terminus.ui.Ui;
+
+public class GoCommand extends WorkspaceCommand {
+
+    public static final String SPACE_DELIMITER = "\\s+";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/GoCommand.java:14]: I guess we could consider moving this to CommonFormat.java

File: src/main/java/terminus/command/ModuleCommand.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    /**
+     * Returns the format for the command.
+     *
+     * @return The String object holding the appropriate format for the command.
+     */
+    @Override
+    public String getFormat() {
+        return "module";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/ModuleCommand.java]: Could consider using a constant for this and refactor it to CommonFormat.java

File: src/main/java/terminus/command/ModuleCommand.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    /**
+     * Returns the description for the command.
+     *
+     * @return The String object containing the description for this command.
+     */
+    @Override
+    public String getHelpMessage() {
+        return "Move to the module workspace";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/ModuleCommand.java]: Could consider using a constant and refactor this to Messages.java

File: src/main/java/terminus/command/module/AddModuleCommand.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import terminus.common.Messages;
+import terminus.common.TerminusLogger;
+import terminus.exception.InvalidArgumentException;
+import terminus.exception.InvalidCommandException;
+import terminus.module.ModuleManager;
+import terminus.ui.Ui;
+
+public class AddModuleCommand extends Command {
+
+    public static final String SPACE_DELIMITER = "\\S+";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/module/AddModuleCommand.java:16]: I guess we could consider moving this to CommonFormat.java

File: src/main/java/terminus/command/module/AddModuleCommand.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * Returns the description for the command.
+     *
+     * @return The String object containing the description for this command.
+     */
+    @Override
+    public String getHelpMessage() {
+        return "Adds a module";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/module/AddModuleCommand.java:37]: Could consider using a constant and move it to Messages.java

File: src/main/java/terminus/command/module/AddModuleCommand.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    private String moduleName;
+
+    /**
+     * Returns the format for the command.
+     *
+     * @return The String object holding the appropriate format for the command.
+     */
+    @Override
+    public String getFormat() {
+        return "add \"<module name>\"";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/module/AddModuleCommand.java:27]: Could consider using a constant and refactor this to CommonFormat.java

File: src/main/java/terminus/command/module/ViewModuleCommand.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+public class ViewModuleCommand extends Command {
+
+    /**
+     * Returns the format for the command.
+     *
+     * @return The String object holding the appropriate format for the command.
+     */
+    @Override
+    public String getFormat() {
+        return "view";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/module/ViewModuleCommand.java:22]: Could consider using a constant and refactor this to CommonFormat.java

File: src/main/java/terminus/command/module/ViewModuleCommand.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    /**
+     * Returns the description for the command.
+     *
+     * @return The String object containing the description for this command.
+     */
+    @Override
+    public String getHelpMessage() {
+        return "View all modules available";
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/module/ViewModuleCommand.java:32]: Could consider using a constant and refactor this to Messages.java

File: src/main/java/terminus/command/content/DeleteCommand.java [hunk 11]
```diff
@@ -78,6 +78,7 @@ public void parseArguments(String arguments) throws InvalidArgumentException {
      */
     @Override
     public CommandResult execute(Ui ui, ModuleManager moduleManager) throws InvalidArgumentException {
+        assert getModuleName() != null;
```
Comment (LouisLouis19) [inline src/main/java/terminus/command/content/DeleteCommand.java:81]: good use of assertion!

**Review by LouisLouis19 — APPROVED** [review]
Looks good overall! Well done!

**Pull request merged**